File: 2016_11_05/shellsort/shellsort.lua
#!/usr/bin/lua

function shell(a)
	local g = math.floor(#a/2)
	while g > 0 do
		for i = g,#a do
			local t = a[i]
			local k = i
			for j = i,g+1,-g do
				if t < a[j-g] then
					a[j] = a[j-g]
					k = k - g
				else
					break
				end
			end
			a[k] = t
		end
		g = math.floor(g/2)
	end
end

local size = tonumber(arg[1])
local test = {}
for i = 1,size do
	test[i] = i
end
math.randomseed(os.time())
for i = 1,size do
	local r = math.random(i, size)
	test[i], test[r] = test[r], test[i]
end

print(table.concat(test, " "))
shell(test)
print(table.concat(test, " "))

File: 2016_11_05/shellsort/shellsort.py
#!/usr/bin/python3

import random, sys, time

def shell(a):
	g = len(a) // 2
	while g > 0:
		for i in range(g, len(a)):
			t = a[i]
			k = i
			for j in range(i,0,-g):
				if t < a[j-g]:
					a[j] = a[j-g]
					k -= g
				else:
					break
			a[k] = t
		g //= 2

size = int(sys.argv[1])
test = list(range(size))
random.shuffle(test)

print(test)
begin = time.time()
shell(test)
end = time.time()
print(test)
